### ISSUE TITLE
Pyright fails to bind function arguments to `*tuple[...]`

### ISSUE THREAD

**Issue by LeeeeT** [microsoft/pyright#7588]
I'm trying to implement a `curry` function. Here's my implementation:

```python
from collections.abc import Callable

def curry[First, *Rest, Result](function: Callable[[First, *Rest], Result]) -> Callable[[*Rest], Callable[[First], Result]]:
    return lambda *rest: lambda first: function(first, *rest)
```

Here's how it works:

```python
def f(a: str, b: int, c: bool) -> None:
    pass

applied_once = curry(f)
applied_twice = curry(applied_once)

reveal_type(applied_once)
reveal_type(applied_twice)
```

```
file.py:12:13 - information: Type of "applied_once" is "(int, bool) -> ((str) -> None)"
file.py:13:13 - information: Type of "applied_twice" is "(bool) -> ((int) -> ((str) -> None))"
```

So, it works fine when creating a temporary variable for each application. But if you directly apply it two times, pyright throws an error:

```python
applied_twice = curry(curry(f))
```

```
file.py:9:23 - error: Argument of type "(*tuple[*tuple[*tuple[*tuple[*tuple[Union[*Rest@curry]]]]]]) -> ((str) -> None)" cannot be assigned to parameter "function" of type "(First@curry, *Rest@curry) -> Result@curry" in function "curry"
  Type "(*tuple[*tuple[*tuple[*tuple[*tuple[Union[*Rest@curry]]]]]]) -> ((str) -> None)" cannot be assigned to type "(Union[*Rest@curry], *Union[*Rest@curry]) -> ((str) -> None)"
    Parameter 1: type "Union[*Rest@curry]" cannot be assigned to type "*tuple[*tuple[*tuple[*tuple[*tuple[Union[*Rest@curry]]]]]]"
      "*tuple[*tuple[*tuple[*tuple[Union[*Rest@curry]]]]]" is incompatible with "*tuple[*tuple[*tuple[*tuple[*tuple[Union[*Rest@curry]]]]]]"
        Tuple entry 1 is incorrect type              
          "*tuple[*tuple[*tuple[Union[*Rest@curry]]]]" is incompatible with "*tuple[*tuple[*tuple[*tuple[Union[*Rest@curry]]]]]"
file.py:9:29 - error: Argument of type "(a: str, b: int, c: bool) -> None" cannot be assigned to parameter "function" of type "(First(1)@curry, *Rest(1)@curry) -> Result(1)@curry" in function "curry"
  Type "(a: str, b: int, c: bool) -> None" cannot be assigned to type "(str, *tuple[*tuple[Union[*Rest@curry]]]) -> None"
    Parameter 2: type "*tuple[*tuple[Union[*Rest@curry]]]" cannot be assigned to type "int"
      "*tuple[*tuple[Union[*Rest@curry]]]" is incompatible with "int"
```

**Comment by erictraut** [comment]
Thanks for the bug report. I'm able to repro, and I agree this is a bug. Interestingly, if you fall back to the older (pre PEP 695) way of defining the type variables, it works without error. You can use that as a temporary workaround. I'll investigate further.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.357, which I just published.
